dune/gdt/algorithms/newton.hh

```diff
@@ -0,0 +1,144 @@
+             << ", initial_guess_vector.sup_norm()=" << initial_guess_vector.sup_norm() << ", param=" << print(param)
+             << ", opts " << print(opts, {{"oneline", "true"}}) << ")" << std::endl;
+  // some preparations
+  const auto& rhs = rhs_vector.as_imp();
+  auto& initial_guess = initial_guess_vector.as_imp();
+  auto residual_op = lhs_operator - rhs;
+  auto residual = rhs.copy();
+  auto update = initial_guess.copy();
+  auto candidate = initial_guess.copy();
+  const auto default_opts = default_newton_solve_options();
+  const auto precision = opts.get("precision", default_opts.get<double>("precision"));
+  const auto max_iter = opts.get("max_iter", default_opts.get<size_t>("max_iter"));
+  const auto max_dampening_iter = opts.get("max_dampening_iter", default_opts.get<size_t>("max_dampening_iter"));
+  // one matrix for all jacobians
+  auto jacobian_op = residual_op.empty_jacobian_op();
+  auto jacobian_solver = XT::LA::make_solver(jacobian_op.matrix());
+  LOG(info) << "solving system by dampened newton:" << std::endl;
+  size_t l = 0;
+  Timer timer;
+  while (true) {
+    const auto prefix = "      " + XT::Common::whitespaceify(l);
+    timer.reset();
+    LOG(info) << "l = " << l << ": computing residual ... " << std::flush;
+    residual_op.apply(initial_guess, residual, param);
+    auto res = residual.l2_norm();
+    LOG(info) << "took " << timer.elapsed() << "s, |residual|_l2 = " << res << std::endl;
+    if (res < precision) {
+      LOG(info) << prefix << "residual below tolerance, succeeded!" << std::endl;
+      break;
+    }
+    DUNE_THROW_IF(l >= max_iter,
+                  Exceptions::newton_error,
+                  "max iterations reached!\n|residual|_l2 = " << res << "\nopts:\n"
+                                                              << opts);
+    LOG(info) << prefix << "computing jacobi matrix ... " << std::flush;
+    timer.reset();
+    jacobian_op.matrix() *= 0.;
+    residual_op.jacobian(initial_guess, jacobian_op, {{"type", residual_op.jacobian_options().at(0)}}, param);
+    jacobian_op.assemble(/*use_tbb=*/true);
+    LOG(info) << "took " << timer.elapsed() << "s\n" << prefix << "solving for defect ... " << std::flush;
+    timer.reset();
+    residual *= -1.;
+    update = initial_guess; // <- initial guess for the linear solver
+    bool linear_solve_succeeded = false;
+    std::vector<std::string> tried_linear_solvers;
+    for (const auto& linear_solver_type : jacobian_solver.types()) {
+      try {
+        tried_linear_solvers.push_back(linear_solver_type);
+        jacobian_solver.apply(
+            residual, update, {{"type", linear_solver_type}, {"precision", XT::Common::to_string(0.1 * precision)}});
+        linear_solve_succeeded = true;
+        break;
+      } catch (const XT::LA::Exceptions::linear_solver_failed&) {
+      }
+    }
+    DUNE_THROW_IF(!linear_solve_succeeded,
```

Comment:
Mhm, but this is here for a reason, Felix from a year ago could have told you! :)

For arbitrary nonlinear operators, it happens quite often that systems are barely solvable and that the choice of the linear solver might really make a difference. How should we proceed then?

We could add an option to the newton cfg about the linear solver, where something like `"auto"` is the current behavior. It's not that easy however, since we would also like to be able to pass the linear solver options as a sub cfg in the newton cfg. If the user selects a single linear solver this is easy.

However, a realistic scenario would be to specify my favorite direct solver and iterative solvers, say `"superlu, amg"`, but how do I pass the respective opts as sub cfgs to the newton?